ui/src/chat/ChatScroller/ChatScroller.tsx

```diff
@@ -129,37 +167,53 @@ export default function ChatScroller({
     [messages, keys, whom, brief, chatInfo, prefixedElement]
   );
 
+  const TopLoader = useMemo(
+    () => <Loader show={fetching === 'top'} />,
+    [fetching]
+  );
+  const BottomLoader = useMemo(
+    () => <Loader show={fetching === 'bottom'} />,
+    [fetching]
+  );
+
   const fetchMessages = useCallback(
     async (newer: boolean) => {
+      setFetching(newer ? 'bottom' : 'top');
+
       if (newer) {
-        return useChatState
+        await useChatState
           .getState()
           .fetchNewer(whom, MESSAGE_FETCH_PAGE_SIZE.toString());
+      } else {
+        await useChatState
+          .getState()
+          .fetchOlder(whom, MESSAGE_FETCH_PAGE_SIZE.toString());
       }
 
-      return useChatState
-        .getState()
-        .fetchOlder(whom, MESSAGE_FETCH_PAGE_SIZE.toString());
+      setFetching('initial');
     },
     [whom]
   );
 
   return (
-    <div className="relative h-full flex-1">
-      {messages.size > 0 ? (
-        <ChatWritScroller
-          key={whom}
-          origin="bottom"
-          style={{ height: '100%', padding: '0.5rem' }}
-          data={mess}
-          size={mess.size}
-          pendingSize={0} // TODO
-          scrollTo={scrollTo}
-          averageHeight={48}
-          renderer={renderer}
-          loadRows={fetchMessages}
-        />
-      ) : null}
+    <div className="relative h-full flex-1 p-2">
+      <Virtuoso
+        {...indexData}
+        followOutput
+        alignToBottom
+        className="h-full"
+        atBottomThreshold={250}
+        atTopThreshold={250}
```

Comment:
nit: would be nice to have a constant so we can tune this somewhere instead of inline magic numbers. alternatively could be calculated dynamically, e.g.  *x* number of messages * average message height

also, did you happen to play with the `overscan` prop to see if it affected the responsiveness and performance?

Reply:
No I didn't but might be worth it. It seemed pretty snappy as is, but could maybe be smoother.